### ISSUE TITLE
Enum to string comparison not being flagged

### ISSUE THREAD

**Issue by foges** [microsoft/pyright#5074]
**Describe the bug**
Comparing an enum to a string literal in Python will always be False. Pyright fails to flag this when the string is a member variable.

**To Reproduce**
See code below.

**Expected behavior**
Expected an error of the form `Condition will always evaluate to False since the types "Literal[..]" and "Literal[..]" have no overlap`

**Screenshots or Code**
```python
from enum import Enum 

class Alpha(Enum):
    A = "a"

class Beta:
    b: str
    def __init__(self):
        self.b = "b"

b = Beta()

c = b.b == Alpha.A   # <- Error is not flagged here
d = "b" == Alpha.A   # <- Error is flagged here as expected
```

<img width="572" alt="image" src="https://user-images.githubusercontent.com/1754288/236527600-2e75d0b5-66f9-4fb5-bb37-7508b2fdad08.png">

**VS Code extension or command-line**
Using VS Code
-   Pyright version: 1.1.305
-   OS and version: Mac OS 12.0.1
-   Python version (& distribution if applicable, e.g. Anaconda): 3.10

**Comment by erictraut** [comment]
I presume that you have the `reportUnnecessaryComparison` check enabled. (Note: this check is not enabled by default unless you are using strict mode.)

Pyright is working as designed here. The `reportUnnecessaryComparison` logic is necessary conservative in what it reports as an error to avoid false positives. In Python, any two objects can potentially be compared because the `__eq__` and `__ne__` operators can be overloaded. Most classes that implement these overloads annotate them such that they accept any `object` as an input parameter. For example, the `str` class implements an `__eq__` override that is typed as `def __eq__(self, __value: object) -> bool: ...`. That means pyright must assume that any value of type `object` is accepted as an input.

The reason that pyright is able to validate that `"b" == Alpha.A` is always false is because both the LHS and RHS operands are literals, and the results of the `__eq__` operation are known for all literal types.

**Comment by foges** [comment]
Hi @erictraut. Thanks for getting back to me. That gave me an idea that I can implement a custom `__eq__` for the Enum

```
class Alpha(Enum):
    A = "a"

   def __eq__(self, value: object) -> bool:
        if isinstance(value, Alpha):
            return super().__eq__(value)
        raise NotImplementedError
```

That will at least give me a runtime error even if pyright doesn't catch it.
